lib/engine/auto_router.rb

```diff
@@ -28,21 +31,63 @@ def compute(corporation, **opts)
         ]
       end
 
+      # Add a per-node timeout to ensure we don't spend ALL time on first node's paths in huge maps,
+      # which can cause all train routes to conflict with each other
+      node_timeout = path_timeout / nodes.size
+
       now = Time.now
 
       skip_paths = static.flat_map(&:paths).to_h { |path| [path, true] }
 
+      train_routes = Hash.new { |h, k| h[k] = [] }    # map of train to route list
+      path_abort = Hash.new { |h, k| h[k] = false }   # each train has opportunity to abort a branch of the path walk tree
+      route_counter = Hash.new { |h, k| h[k] = 0 }
+
+      trains.each do |train|
```

Comment:
there's no point to do this, you already created hashes on line 43 with default values...